linalg/sparsemat.hpp

```diff
@@ -118,6 +145,12 @@ class SparseMatrix : public AbstractSparseMatrix
    /// Create a SparseMatrix with diagonal @a v, i.e. A = Diag(v)
    SparseMatrix(const Vector & v);
 
+   // Initialize CuSparse
+   void InitCuSparse();
+
+   // Runtime option to use CuSparse
+   // Only valid when using a CUDA backend
+   void UseCuSparse(bool _useCuSparse = true) { useCuSparse = _useCuSparse;}
```

Comment:
I'm not a fan of this method from a design point of view, `UseCuSparse` is not a property that always exists, but this method always does. I would `#ifdef` guard the `bool useCuSparse{true};` and this method. Is the purpose of this method to allow users to choose for each `SparseMatrix` if it should use `CuSparse` or the native implementation? This would sound like an unlikely use case, is there big performance differences that would justify to ever use the native algorithm over `CuSparse`?

Reply:
@YohannDudouit The purpose is to allow the choice between cuSPARSE and the native implementation. It is unclear whether there are significant performance differences between the two, but @artv3 referenced a paper that suggests in some cases the native version may be faster. Another issue is that cuSPARSE allocates some extra memory on the GPU, and it is unclear how much. Considering the uncertainty about performance and memory usage, it seems good to allow the choice between versions at least until more testing is done.

Reply:
I think this is unrealistic to think that users will be profiling each `SparseMatrix` `AddMult` with and without `CuSparse`, this is clearly early optimization to me, this interface will probably end up never being used. I think this interface is missing the point, which is to activate/deactivate __globally__ `CuSparse`, currently you would have to go everywhere in your code where you use a `SparseMatrix` and call `UseCuSparse(false)`, I don't see any use currently for such a feature, and this doesn't really help users to deactivate `CuSparse`.

Reply:
@YohannDudouit brings up a good point about adding the global capability. Is there a good place to add that?

Reply:
@artv3 You could use a `static` attribute inside `SparseMatrix` which defines the default behavior, then you define a static method on `SparseMatrix` to change the default behavior. This could also be done at compilation time like you suggested (which I tend to like more).

@dylan-copeland I think it would be better to do the testing before adding the method, I don't really like the idea to add a "maybe useful" method, that's confusing.

@artv3 @dylan-copeland @v-dobrev If you think that what we need is a method for each sparse matrix I'm okay with it.

Reply:
@YohannDudouit Ideally, I agree that testing should be done before adding a feature, but this case is complicated and may require different users to test different `SparseMatrix` cases. Even if cuSPARSE is the best for every possible matrix, users may want to verify that for their tests. I think it is good to allow the user options to globally or locally switch between the native version, cuSPARSE, and other libraries to be added in the future. You made a good point about local switching being too fine in granularity, but it is hard to foresee what kinds of tests a user might want to do. I would err on the side of caution by giving the user flexibility to test all available solvers for any matrix.

Reply:
Okay, let's keep this for now.